apps/consulting/posts/review-functorch-contribution.md

```diff
@@ -0,0 +1,243 @@
+---
+title: 'Review: Functorch Contribution'
+published: September 22, 2023
+author: kshiteej-kalambarkar
+description: '`functorch` (now `torch.func`) is a PyTorch module to provide composable transforms like JAX. It provides transforms like `grad`, `vmap`, `vjp`, etc. to allow user to easily compute Jacobian vector products and vectorize the function. And being composable, one can compute per-sample gradients simply by using `vmap(grad(model))`.'
+---
+
+<base target="_blank" />
+
+`functorch` (now `torch.func`) is a PyTorch module to provide composable transforms like JAX. It provides transforms like `grad`, `vmap`, `vjp`, etc. to allow user to easily compute Jacobian vector products and vectorize the function. And being composable, one can compute per-sample gradients simply by using `vmap(grad(model))`.
+
+Below are some of the things that I got to work on:
+
+## Adding Batching Rule for `vmap`
+`vmap` is a transform which takes a function `func` that runs on a single datapoint and returns a function which can handle a batch of data effectively vectorizing it. Semantically, it runs a `for`-loop over all data points and stacks all the results together (but does it more optimally than the for-loop version.)
+
+Example:
+```python
+import torch
+
+# Written to handle only single sample.
+# Calling it with batched input, will
+# result in incorrect output.
+def my_simple_model(input, weight, bias):
+    return input.sum(0) @ weight + bias
```

Comment:
Consider rather than using he odd `input.sum(0)` construction, doing the more realistic `torch.mv(weight, input)`, for example.